### PR TITLE
cgi-bin/ipp-var.c: Use "guest" user for Move Job when no REMOTE_USER given

### PR DESCRIPTION
**Pull request by bmasonrh**
`cgi-bin/do_job_op()` sets `user` to "`guest`" if `REMOTE_USER` isn't set.  However, `cgiMoveJob()` (in `cgi-bin/ipp-var.c`) returns 401/Unauthorized if `REMOTE_USER` isn't set.  This inconsistent behavior leads to the following issue:

If an administrator wants to allow unfettered access to allow unauthenticated users to perform administration functions, they can do this by removing `Require user` entries from `cupsd.conf`.  (We can debate the wisdom of this sort of configuration, but I have a customer who wants to do this.)  This works for everything but the "Move Job" operation in the "Jobs" page.  Because `cgiMoveJob()` returns 401, it is impossible to move the job using the Web UI.  This patch fixes that.

With the patch applied and `Require user` entries removed from `cupsd.conf`, unauthenticated users can move print jobs using the Web UI.  

With the patch applied and the default `cupsd.conf`, authentication occurs later in the process of moving a print job, through the following sequence of events:
1. Go to the "Jobs" page (`http://<lhost>:631/jobs/`)
2. Press "Move Job" -- the page with the destination list is shown
3. Select a destination and press "Move Job" -- the "Upgrade Required" page is shown and the browser is redirected back to the "Jobs" page (`https://<host>:631/jobs/`.
4. Press "Move Job" again -- page with destination list is shown
5. Select a destination and press "Move Job" -- browser prompts for authentication
6. Enter credentials -- job is moved

Although this path seems a bit circuitous, it's not dissimilar to the process needed to modify a printer, where navigating to a printer page (`http://<host>/printers/<printer>`) and selecting "Modify Printer" from the "Administration" drop-down results in an "Upgrade Required" redirect to the "Administration" page (`https://<host>:631/admin/`)

### CODE REVIEW

**Comment by zdohnal** [comment]
Hi Bryan,

thank you for the PR!

I'm able to reproduce the issue with our latest upstream revision by steps (being in CUPS github repository root):

```
(0. get CUPS development files installed)
1. $ ./configure --enable-webif
2. $ make
3. $ sudo make install
4. remove 'Require user' lines from /etc/cups/cupsd.conf
5. add 'FileDevice Yes' into /etc/cups/cups-files.conf
6. start cupsd (in my case `$ sudo systemctl start cups`)
7. $ lpadmin -p test1 -v file:/tmp/ps1 -m drv:///sample.drv/generic.ppd -E
8. $ lpadmin -p test2 -v file:/tmp/ps2 -m drv:///sample.drv/generic.ppd -E
9. $ cupsdisable test1
10. $ lp -d test1 examples/testfile.ps
```

Then I went to CUPS web UI at localhost:631, jobs panel, where the job is showed being stuck in test1 queue -> clicked on 'Move job' and the UI wanted a credentials from me, even if the lines 'Require user' are removed from cupsd.conf.

Ad the patch, I'm not sure whether we want to enable this - if it is enabled, anyone with access to web ui can redirect any held jobs to the print queue of their liking, especially with combination anyone can create a queue via web interface...

**Comment by bmasonrh** [comment]
> Ad the patch, I'm not sure whether we want to enable this - if it is enabled, anyone with access to web ui can redirect any held jobs to the print queue of their liking, especially with combination anyone can create a queue via web interface...

That's the point, though.  If I've removed all of the `Require user` lines, then I have explicitly configured CUPS to allow **anyone** to be able to do **anything**.  But without the patch they can't do this one operation.

**Comment by zdohnal** [comment]
> That's the point, though. If I've removed all of the `Require user` lines, then I have explicitly configured CUPS to allow **anyone** to be able to do **anything**. But without the patch they can't do this one operation.

What I mean is I'm not sure whether the core of issue is we don't have a last line of defense of all other operations except for moving the job, or if the moving of the job is the only one which still wants the creds even against the conf says otherwise.

IIRC some applications have such a last line of defense for the functionality, which can lead into critical malfunction or leak of important data. I don't know yet which approach we take in CUPS regarding this or if moving the job is critical enough to prevent an unauthenticated access even if the conf says otherwise.

@michaelrsweet wdyt?

**Comment by michaelrsweet** [comment]
@zdohnal As long as we retain the same level of security as, say, Cancel-Job, then this should be fine. But we want to make sure that *is* the case (of course)...

**Comment by zdohnal** [comment]
@michaelrsweet since Cancel-Job is allowed for unauthenticated user (once Require directives are removed from cupsd.conf), and we want to retain the same level of security, then the PR is okay from the logic side.
However the code style should be updated, I'll do it after the merge.

**Review by zdohnal — APPROVED** [review]
I'll fix the code style after the merge - the semantic is the same as the other operations after applying the patch, so LGTM.

**Comment by zdohnal** [comment]
@michaelrsweet IMO the PR is correct, do you agree with the reasons I've provided?

File: cgi-bin/ipp-var.c [hunk 1]
```diff
@@ -272,8 +272,7 @@ cgiMoveJobs(http_t     *http,		/* I - Connection to server */
 
   if ((user = getenv("REMOTE_USER")) == NULL)
   {
-    puts("Status: 401\n");
-    exit(0);
+    user = "guest";
```
[the remark this answers is not in the document]
Comment (bmasonrh) [reply cgi-bin/ipp-var.c:275]: OK.  Missed that (didn't realize I had reduced the scope to one line).  Will fix it so you can merge it easily.

**Comment by zdohnal** [comment]
Thank you for the review @michaelrsweet !

Thank you for the update and the PR, @bmasonrh !

**Pull request merged**